doc/specs/#13666 - VT Sequence for Screen Reader Control.md

```diff
@@ -0,0 +1,105 @@
+---
+author: Carlos Zamora @carlos-zamora
+created on: 2022-11-03
+last updated: 2022-11-04
+issue id: 13666
+---
+
+# VT Sequence for Screen Reader Control
+
+## Abstract
+
+Now that Windows Terminal has a UI Automation implementation that can be leveraged by screen readers, the content of the buffer can be explored and engaged with via screen readers. Additionally, the introduction of UIA notifications opened the possibility for screen readers to directly speak content output from the terminal application. However, not all command-line application scenarios are necessarily accessible as they are generally limited to moving the cursor and writing output to the cursor's position. As such, command-line applications require a way to directly interact with the screen reader as they know the best way to represent their content.
+
+## Inspiration
+
+In 2022, a community member reported a bug on Windows Terminal that the GitHub CLI command `gh pr create` would not be read out properly[^1]. The problem was caused by GitHub CLI rewriting a prompt to highlight a selected option using a '>' prefix. This resulted in screen readers re-reading the entire prompt instead of only the selected option because Windows Terminal dispatches UIA notifications for output text.
+
+In 2019, Daniel Imms proposed a VT sequence to control screen readers directly from the command-line application[^2]. In summary, this VT sequence allows command-line applications to (1) stop reading new output temporarily, (2) override what text is read by the screen reader, and (3) supply additional text to be ready by the screen reader. As presented in the spec from ther Terminal working group:
+> `OSC Ps ; Pt ST`
+> - `Ps = 2 0 0` -> Stop announcing incoming data to screen reader, `Pt` is an optional string that will be announced immediately. The screen reader will resume announcing incoming data if any key is pressed.
+> - `Ps = 2 0 1` -> Resume announcing incoming data to screen reader, `Pt` is an optional string that will be announced immediately.
+> - `Ps = 2 0 2` -> Announce `Pt` immediately to the screen reader.
+> Note that the reason any key press will force the screen reader to announce again is to prevent situations where applications are terminated while the screen reader is not announcing or where applications are misbehaving.
```

Comment:
This is a bit of a nit pick, but it's worth noting that OSC numbers are a finite resource, so a single OSC number for this would be preferable to three. For example, something like `OSC 200 ; Ps ; Pt`, where `Ps` differentiates between stop/resume/announce. Also makes it a little easier to extend.

My main concern, though, is how this is going to propagate over conpty. Is the idea to just pass it through and hope for the best? What happens if conpty later refreshes part of the display that was originally output with "stop announcing"? Would we then need to rewrap that content with these sequences?

Because if that is something conpty needs to account for, we may be better off with a simple attribute-like sequence similar to [`DECSCA`](https://vt100.net/docs/vt510-rm/DECSCA.html), which could be recorded in the buffer, and then forwarded over conpty as part the regular repaint. The "immediate announce" strings would still then need a separate sequence (and that probably could be passed through directly).

I don't know. Just thinking out loud. This may be something you need to prototype and try out before you lock down the exact protocol you're going to use.

Reply:
> This is a bit of a nit pick, but it's worth noting that OSC numbers are a finite resource, so a single OSC number for this would be preferable to three. For example, something like `OSC 200 ; Ps ; Pt`, where `Ps` differentiates between stop/resume/announce. Also makes it a little easier to extend.

Oh I like that!

> My main concern, though, is how this is going to propagate over conpty. Is the idea to just pass it through and hope for the best? What happens if conpty later refreshes part of the display that was originally output with "stop announcing"? Would we then need to rewrap that content with these sequences?

I always felt like ConPTY recording and then "re-rendering" VT output is quite a bit of a "hack". And because of that, if we ever design a VT sequence, I don't think we should limit ourselves by how hard it'd be to implement inside conhost.
Basically, I'd personally be in favor of whatever optimal / "clean" design we can come up with, unhindered by any design complexities that only we would have to suffer, unlike other, existing UNIX terminals. Long-term `VtEngine` might just be replaced entirely with something leaner anyways.

Reply:
> I always felt like ConPTY recording and then "re-rendering" VT output is quite a bit of a "hack". And because of that, if we ever design a VT sequence, I don't think we should limit ourselves by how hard it'd be to implement inside conhost.

I very much agree with this sentiment, but I am concerned we might end up with something we can't actually use. And while I'm still confident we can improve on the current `VtEngine`, that's a bigger task than I had originally thought, and I don't see that happening anytime soon.

That said, this may end up not being that big a deal in practice. I just wanted everyone to be aware of the potential issues here.

Reply:
> be aware of the potential issues here

I'll be sure to mention this in the potential issues part of the spec 😉. I think my stance so far though, is basically the same as my comments on scan mode: https://github.com/microsoft/terminal/blob/60d368590d7a6b59a70de42b09877da2b145dfa1/doc/specs/%2313666%20-%20VT%20Sequence%20for%20Screen%20Reader%20Control.md?plain=1#L58-L69

This sequence would be most useful for text being output rather than leaving a "landmark" stored in the buffer. The "alt text" sequence is probably something we would want to tackle separately, if desired and found to be useful.

That said, I'd be concerned about the scenario where the user switches from the alt buffer back to the main buffer. That probably wouldn't work right. Urgh. I'll have to be sure to test that out. (another one for "potential issues" haha) :/